src/cargo/util/config.rs

```diff
@@ -427,38 +487,67 @@ impl Config {
         }
     }
 
+    // Recommend use `get` if you want a specific type, such as an unsigned value.
+    // Example:  config.get::<Option<u32>>("some.key")?
     pub fn get_i64(&self, key: &str) -> CargoResult<Option<Value<i64>>> {
-        if let Some(v) = self.get_env(key)? {
-            return Ok(Some(v));
-        }
-        match self.get(key)? {
-            Some(CV::Integer(i, path)) => Ok(Some(Value {
-                val: i,
-                definition: Definition::Path(path),
-            })),
-            Some(val) => self.expected("integer", key, val),
-            None => Ok(None),
-        }
+        self.get_integer(&ConfigKey::from_str(key))
+            .map_err(|e| e.into())
     }
 
-    pub fn net_retry(&self) -> CargoResult<i64> {
-        match self.get_i64("net.retry")? {
-            Some(v) => {
-                let value = v.val;
-                if value < 0 {
-                    bail!(
-                        "net.retry must be positive, but found {} in {}",
-                        v.val,
-                        v.definition
-                    )
+    fn get_integer<T>(&self, key: &ConfigKey) -> Result<Option<Value<T>>, ConfigError>
+    where
+        T: FromStr + num_traits::NumCast + num_traits::Bounded + num_traits::Zero + fmt::Display,
```

Comment:
I'm actually a little surprised this is necessary, shouldn't Serde be doing all this under the hood? In that if you deserialize a -1 during a `visit_u32`, I think serde returns a canned error?

Reply:
Yea, you're right.  The error messages aren't quite as nice, but are good enough, and probably not worth the extra complexity.